src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -50,27 +225,296 @@ def execute(self, reqStatus):
+        :return:        number of properly cleaned requests
+        """
+        # NOTE: We iterate destructively through the reqList
+        # NOTE: The Input Cleanup, the Block Level Cleanup and the Archival
+        #       Pipelines are executed sequentially in the above order.
+        #       This way we assure ourselves that we archive only workflows
+        #       that have accomplished the needed cleanup
+
+        clean_num_requests = 0
+        total_num_requests = 0
+
+        # Call the workflow dispatcher:
+        for reqName, req in reqRecords.items():
+            wflow = MSRuleCleanerWflow(reqRecords.pop(reqName))
+            self._dispatchWflow(wflow)
+            msg = "\n----------------------------------------------------------"
+            msg += "\nMSRuleCleanerWflow: %s"
+            msg += "\n----------------------------------------------------------"
+            self.logger.debug(msg, pformat(wflow))
+            total_num_requests += 1
+            if self._checkClean(wflow):
+                clean_num_requests += 1
+
+        # Report the counters:
+        for pline in self.cleanuplines:
+            msg = "%s: Workflows cleaned by pipeline: %s: %d"
+            self.logger.info(msg, self.currThreadIdent, pline.name, self.wfCounters['cleaned'][pline.name])
+        archived_num_requests = self.wfCounters['archived']
+        msg = "%s: Workflows archived: %d"
+        self.logger.info(msg, self.currThreadIdent, self.wfCounters['archived'])
+        return clean_num_requests, total_num_requests, archived_num_requests
+
+    def _dispatchWflow(self, wflow, *args, **kwargs):
+        """
+        A function intended to dispatch a workflow (e.g based on its status)
+        through one or more functional pipelines in case there is some more
+        complicated logic involved in the order we execute them but not just
+        a sequentially
+        """
+        # NOTE: The following dispatch logic is a subject to be changed at any time
+
+        if wflow['RequestType'] == 'Resubmission':
+            pass
+
+        # Clean:
+        elif wflow['RequestStatus'] in ['rejected', 'aborted-completed']:
+            # NOTE: We do not check the ParentageResolved flag for these
+            #       workflows, but we do need to clean output data placement
+            #       rules from the agents for them
+            for pline in self.agentlines:
+                try:
+                    pline.run(wflow)
+                except Exception as ex:
+                    msg = "%s: General error from pipeline. Error: \n%s. "
+                    msg += "\nWill retry again in the next cycle."
+                    self.logger.exception(msg, pline.name, str(ex))
+                    break
+                if wflow['CleanupStatus'][pline.name]:
+                    self.wfCounters['cleaned'][pline.name] += 1
+        elif wflow['RequestStatus'] == 'announced' and wflow['ParentageResolved']:
```

Comment:
Done, But... It may be also considered as a filter and I wondered if it would not be a better idea to move it in `_execute()` where I started applying the rest of the filters with moving the `Resubmission` up. In both cases we should take into account there is a default value for the 'ParentageResoved'  flag set to `True`  in the MSRuleCleaner object creation.  
We should decide this before. Tell me what do you think?

Reply:
It could be done over there too. Maybe even wrapping those 2 checks in an utilitarian method and only convert to MSRuleCleanerWflow object if it passes those constraints...

Reply:
Oh, and this last `elif` statement can likely be an `else` now, right?

Reply:
yes, it can

Reply:
ok ... thinking twice .. I like it more with an `elif`, because it states explicitly on which workflows it is about to be working on.... if we make it an `else`, than we make it too general and everything which does not suffice  the above conditions will fall under this category and will be worked on just like the proper `announced` with 'ParentageResolved'. In the current set of conditions above we are safe, because we expect only those to reach down here, but imagine we decide to put or take out something above later on and forget to take this int account here... then... voile a bug is stalking behind the corner :)